data_management/rocrate.py

```diff
@@ -0,0 +1,661 @@
+
+    """
+    if (
+        software_object.storage_location.public is True
+        and len(str(software_object.storage_location).split(FILE)) > 1
+    ):
+        source_loc = str(software_object.storage_location).split(FILE)[1]
+        dest_path = f"inputs/{software_type}/{source_loc.split('/')[-1]}"
+
+    else:
+        source_loc = (
+            f"{registry_url}api/storage_location/{software_object.storage_location.id}"
+        )
+        dest_path = None
+
+    crate_software_object = crate.add_file(
+        source_loc,
+        dest_path=dest_path,
+        properties={
+            RO_TYPE: ["File", "SoftwareSourceCode"],
+            "name": str(software_object.storage_location).split("/")[-1],
+        },
+    )
+
+    if software_object.description is not None:
+        crate_software_object["description"] = software_object.description
+
+    if (
+        software_object.file_type is not None
+        and software_object.file_type.extension is not None
+    ):
+        crate_software_object["encodingFormat"] = _get_mime_type(
+            software_object.file_type.extension
+        )
+
+    if software_object.storage_location.hash is not None:
+        crate_software_object["hash"] = software_object.storage_location.hash
+
+    _add_licenses(crate, crate_software_object, software_object, registry_url)
+
+    _add_authors(
+        software_object.authors.all(),
+        crate,
+        crate_software_object,
+        registry_url,
+    )
+
+    return crate_software_object
+
+
+def generate_ro_crate_from_cr(code_run, request):
+    """
+    Crate an RO Crate based around the code run.
+
+    @param code_run: a code_run from the CodeRun table
+
+    @return the RO Crate object
+
+    """
+    mimetypes.init()
+    registry_url = request.build_absolute_uri("/")
+
+    crate = ROCrate()
+    crate.publisher = "FAIR Data Pipeline"
+    crate.datePublished = datetime.now().isoformat()
+    crate.name = f"RO Crate for code run {code_run.id}"
+
+    # add the licenses from each of the data products to the ROCrate
+    for output in code_run.outputs.all():
+        _add_licenses(crate, crate, output.object, registry_url)
+    if crate.license is None:
+        crate_license = _get_default_license(crate)
+        crate.add(crate_license)
+        crate.license = crate_license
+
+    _add_metadata_license(crate)
+
+    instruments = []
+    crate_data_product = None
+
+    # add the code run
+    crate_code_run = _get_code_run(crate_data_product, crate, code_run, registry_url)
+
+    # add the code repo release
+    if code_run.code_repo is not None:
+        code_release = _get_code_repo_release(crate, code_run.code_repo, registry_url)
+        instruments.append(code_release)
+
+    # add the model config
+    if code_run.model_config is not None:
+        model_config = _get_software(
+            crate, code_run.model_config, registry_url, "model_config"
+        )
+        instruments.append(model_config)
```

Comment:
The only instrument should be the software application whose run is represented by the CreateAction. Anything else that's needed for it to run should be listed in "object".

Reply:
The submission script and config file have been moved to the `object` in https://github.com/FAIRDataPipeline/data-registry/commit/64d9b6daada70227791f778ab7ee67b91ee84d7a